File: task_002_google_drive_get_query/evaluation/run_evaluation.py
#!/usr/bin/env python3
"""
Evaluation runner that generates structured reports.

Run with:
    docker compose run --rm app python evaluation/run_evaluation.py [options]

Options:
    --users N           Number of users to seed (default: 21)
    --folders N         Number of folders to seed (default: 100)
    --files-per-folder N  Files per folder (default: 50)
    --iterations N      Iterations per test (default: 5)
    --output-dir DIR    Output directory (default: evaluation/)
"""
import os
import sys
import json
import uuid
import subprocess
import platform
from datetime import datetime
from pathlib import Path
import argparse

# Add project root to path
sys.path.insert(0, str(Path(__file__).parent.parent))


def get_git_commit():
    """Get current git commit hash."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--short", "HEAD"],
            capture_output=True, text=True, timeout=5
        )
        return result.stdout.strip() if result.returncode == 0 else "unknown"
    except Exception:
        return "unknown"


def get_git_branch():
    """Get current git branch."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True, text=True, timeout=5
        )
        return result.stdout.strip() if result.returncode == 0 else "unknown"
    except Exception:
        return "unknown"


def get_environment_info():
    """Collect environment information."""
    return {
        "python_version": platform.python_version(),
        "platform": platform.platform(),
        "os": platform.system(),
        "os_release": platform.release(),
        "architecture": platform.machine(),
        "docker_image": os.environ.get("DOCKER_IMAGE", "python:3.11-slim"),
        "git_commit": get_git_commit(),
        "git_branch": get_git_branch(),
        "database_url": os.environ.get("DATABASE_URL", "sqlite (in-memory)"),
    }


def run_performance_evaluation(params):
    """
    Run performance evaluation and collect metrics.
    
    Returns dict with metrics from both before and after implementations.
    """
    from repository_before.app import create_app as create_app_before
    from repository_after.app import create_app as create_app_after
    from repository_before import db as db_before
    from tests.utils import (
        clear_database,
        seed_heavy_user_data,
        measure_performance,
    )
    
    db_url = os.environ.get("DATABASE_URL", "sqlite:///evaluation_test.db")
    
    # Initialize before app
    before_app = create_app_before(db_url)
    before_app.config["TESTING"] = True
    
    # Seed data
    session = db_before.SessionLocal()
    try:
        clear_database(session)
        seed_stats = seed_heavy_user_data(
            session, 
            num_folders=params["folders"], 
            num_files_per_folder=params["files_per_folder"]
        )
    finally:
        session.close()
    
    # Initialize after app
    after_app = create_app_after(db_url)
    after_app.config["TESTING"] = True
    
    users_to_test = ["heavy_user", "other_user_0", "other_user_5", "other_user_10"]
    
    before_results = {}
    after_results = {}
    
    # Run measurements
    for user_id in users_to_test:
        with before_app.test_client() as client:
            before_results[user_id] = measure_performance(
                client, user_id, 
                num_iterations=params["iterations"]
            )
        
        with after_app.test_client() as client:
            after_results[user_id] = measure_performance(
                client, user_id, 
                num_iterations=params["iterations"]
            )
    
    # Cleanup
    session = db_before.SessionLocal()
    try:
        clear_database(session)
    finally:
        session.close()
    
    # Calculate aggregate metrics
    metrics = {
        "seed_stats": seed_stats,
        "before": {},
        "after": {},
        "comparison": {}
    }
    
    for user_id in users_to_test:
        b = before_results[user_id]
        a = after_results[user_id]
        
        metrics["before"][user_id] = {
            "avg_ms": b["avg_ms"],
            "min_ms": b["min_ms"],
            "max_ms": b["max_ms"],
            "folders": b["folders"],
            "files": b["files"],
        }
        
        metrics["after"][user_id] = {
            "avg_ms": a["avg_ms"],
            "min_ms": a["min_ms"],
            "max_ms": a["max_ms"],
            "folders": a["folders"],
            "files": a["files"],
        }
        
        speedup = b["avg_ms"] / a["avg_ms"] if a["avg_ms"] > 0 else 0
        improvement_pct = ((b["avg_ms"] - a["avg_ms"]) / b["avg_ms"]) * 100 if b["avg_ms"] > 0 else 0
        
        metrics["comparison"][user_id] = {
            "speedup": round(speedup, 2),
            "improvement_pct": round(improvement_pct, 1),
        }
    
    # Overall averages
    before_avg = sum(m["avg_ms"] for m in metrics["before"].values()) / len(users_to_test)
    after_avg = sum(m["avg_ms"] for m in metrics["after"].values()) / len(users_to_test)
    overall_speedup = before_avg / after_avg if after_avg > 0 else 0
    
    metrics["summary"] = {
        "before_avg_ms": round(before_avg, 2),
        "after_avg_ms": round(after_avg, 2),
        "overall_speedup": round(overall_speedup, 2),
        "overall_improvement_pct": round(((before_avg - after_avg) / before_avg) * 100 if before_avg > 0 else 0, 1),
    }
    
    return metrics


def generate_markdown_report(report_data):
    """Generate a human-readable markdown report."""
    env = report_data["environment"]
    params = report_data["parameters"]
    metrics = report_data["metrics"]
    
    db_url_display = env['database_url']
    if len(db_url_display) > 50:
        db_url_display = db_url_display[:50] + "..."
    
    md = f"""# Performance Evaluation Report

**Run ID:** `{report_data['run_id']}`  
**Started:** {report_data['started_at']}  
**Finished:** {report_data['finished_at']}  
**Duration:** {report_data['duration_seconds']:.2f} seconds

---

## Environment

| Property | Value |
|----------|-------|
| Python Version | {env['python_version']} |
| Platform | {env['platform']} |
| OS | {env['os']} {env['os_release']} |
| Architecture | {env['architecture']} |
| Docker Image | {env['docker_image']} |
| Git Commit | `{env['git_commit']}` |
| Git Branch | `{env['git_branch']}` |
| Database | {db_url_display} |

---

## Parameters

| Parameter | Value |
|-----------|-------|
| Users | {params['users']} |
| Folders | {params['folders']} |
| Files per Folder | {params['files_per_folder']} |
| Total Files | {params['folders'] * params['files_per_folder']} |
| Iterations | {params['iterations']} |

---

## Seed Statistics

| Entity | Count |
|--------|-------|
| Users | {metrics['seed_stats']['users']} |
| Folders | {metrics['seed_stats']['folders']} |
| Files | {metrics['seed_stats']['files']} |
| Permissions | {metrics['seed_stats']['permissions']} |

---

## Summary Results

| Metric | Before (Naive) | After (Optimized) | Improvement |
|--------|----------------|-------------------|-------------|
| Average Response | {metrics['summary']['before_avg_ms']:.2f} ms | {metrics['summary']['after_avg_ms']:.2f} ms | **{metrics['summary']['overall_speedup']:.2f}x faster** |
| Improvement | - | - | {metrics['summary']['overall_improvement_pct']:.1f}% |

---

## Detailed Results by User

| User | Before (ms) | After (ms) | Speedup | Folders | Files |
|------|-------------|------------|---------|---------|-------|
"""
    
    for user_id in metrics["before"]:
        b = metrics["before"][user_id]
        a = metrics["after"][user_id]
        c = metrics["comparison"][user_id]
        md += f"| {user_id} | {b['avg_ms']:.2f} | {a['avg_ms']:.2f} | {c['speedup']:.2f}x | {a['folders']} | {a['files']} |\n"
    
    md += """
---

## Conclusion

"""
    
    if metrics['summary']['overall_speedup'] >= 2.0:
        md += f"✅ **Excellent optimization!** The optimized implementation is **{metrics['summary']['overall_speedup']:.2f}x faster** than the naive version.\n"
    elif metrics['summary']['overall_speedup'] >= 1.0:
        md += f"✅ **Good optimization.** The optimized implementation is **{metrics['summary']['overall_speedup']:.2f}x faster** than the naive version.\n"
    else:
        md += f"⚠️ **Needs investigation.** The optimized implementation is slower than the naive version.\n"
    
    return md


def main():
    parser = argparse.ArgumentParser(description="Run performance evaluation with structured reports")
    parser.add_argument("--users", type=int, default=21, help="Number of users (default: 21)")
    parser.add_argument("--folders", type=int, default=100, help="Number of folders (default: 100)")
    parser.add_argument("--files-per-folder", type=int, default=50, help="Files per folder (default: 50)")
    parser.add_argument("--iterations", type=int, default=5, help="Test iterations (default: 5)")
    parser.add_argument("--output-dir", type=str, default="evaluation", help="Output directory (default: evaluation)")
    args = parser.parse_args()
    
    # Generate run ID and timestamps
    run_id = str(uuid.uuid4())[:8]
    started_at = datetime.now()
    
    # Create output directory with date and timestamp
    date_folder = started_at.strftime("%Y-%m-%d")
    timestamp = started_at.strftime("%H-%M-%S")
    output_path = Path(args.output_dir) / date_folder / timestamp
    output_path.mkdir(parents=True, exist_ok=True)
    
    print(f"{'=' * 60}")
    print(f"PERFORMANCE EVALUATION")
    print(f"{'=' * 60}")
    print(f"Run ID: {run_id}")
    print(f"Started: {started_at.isoformat()}")
    print(f"Output: {output_path}")
    print(f"{'=' * 60}")
    
    parameters = {
        "users": args.users,
        "folders": args.folders,
        "files_per_folder": args.files_per_folder,
        "iterations": args.iterations,
    }
    
    print(f"\nParameters: {json.dumps(parameters, indent=2)}")
    print(f"\nRunning evaluation...")
    
    # Capture stdout for the log
    import io
    from contextlib import redirect_stdout
    
    stdout_capture = io.StringIO()
    
    with redirect_stdout(stdout_capture):
        print(f"Run ID: {run_id}")
        print(f"Started: {started_at.isoformat()}")
        print(f"Parameters: {json.dumps(parameters, indent=2)}")
        print()
        
        try:
            metrics = run_performance_evaluation(parameters)
            
            # Print results to captured stdout
            print(f"\n{'=' * 60}")
            print("EVALUATION RESULTS")
            print(f"{'=' * 60}")
            print(f"\nSeed Statistics: {json.dumps(metrics['seed_stats'], indent=2)}")
            print(f"\nSummary:")
            print(f"  Before (naive) avg:     {metrics['summary']['before_avg_ms']:.2f} ms")
            print(f"  After (optimized) avg:  {metrics['summary']['after_avg_ms']:.2f} ms")
            print(f"  Overall speedup:        {metrics['summary']['overall_speedup']:.2f}x")
            print(f"  Improvement:            {metrics['summary']['overall_improvement_pct']:.1f}%")
            print(f"\nDetailed Results:")
            for user_id in metrics['before']:
                b = metrics['before'][user_id]
                a = metrics['after'][user_id]
                c = metrics['comparison'][user_id]
                print(f"  {user_id}: {b['avg_ms']:.2f}ms -> {a['avg_ms']:.2f}ms ({c['speedup']:.2f}x)")
            
            success = True
            error_message = None
            
        except Exception as e:
            print(f"\nERROR: {str(e)}")
            import traceback
            traceback.print_exc()
            metrics = None
            success = False
            error_message = str(e)
    
    finished_at = datetime.now()
    duration = (finished_at - started_at).total_seconds()
    
    # Build full report
    report_data = {
        "run_id": run_id,
        "started_at": started_at.isoformat(),
        "finished_at": finished_at.isoformat(),
        "duration_seconds": duration,
        "success": success,
        "error": error_message,
        "parameters": parameters,
        "environment": get_environment_info(),
        "metrics": metrics,
    }
    
    # Write report.json
    json_path = output_path / "report.json"
    with open(json_path, "w") as f:
        json.dump(report_data, f, indent=2)
    print(f"\n✅ Saved: {json_path}")
    
    # Write report.md
    if success:
        md_path = output_path / "report.md"
        with open(md_path, "w") as f:
            f.write(generate_markdown_report(report_data))
        print(f"✅ Saved: {md_path}")
    
    # Write stdout.log
    log_path = output_path / "stdout.log"
    with open(log_path, "w") as f:
        f.write(stdout_capture.getvalue())
    print(f"✅ Saved: {log_path}")
    
    print(f"\n{'=' * 60}")
    print(f"EVALUATION COMPLETE")
    print(f"{'=' * 60}")
    print(f"Duration: {duration:.2f}s")
    print(f"Reports saved to: {output_path}")
    
    return 0 if success else 1


if __name__ == "__main__":
    sys.exit(main())

File: BD-RL-002-sw_crm_contact_search/evaluation/run_evaluation.cs
extern alias repository_after;

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using RepoBefore = repository_before;
using RepoAfter = repository_after::repository_before;

internal static class Program
{
    public static async Task<int> Main(string[] args)
    {
        var options = EvalOptions.Parse(args);
        var runId = Guid.NewGuid().ToString("N")[..8];
        var startedAt = DateTime.Now;

        var outputPath = Path.Combine(options.OutputDir, startedAt.ToString("yyyy-MM-dd"), startedAt.ToString("HH-mm-ss"));
        Directory.CreateDirectory(outputPath);

        var originalOut = Console.Out;
        using var teeWriter = new TeeTextWriter(originalOut);
        Console.SetOut(teeWriter);

        Console.WriteLine(new string('=', 60));
        Console.WriteLine("PERFORMANCE EVALUATION (.NET)");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine($"Run ID: {runId}");
        Console.WriteLine($"Started: {startedAt:O}");
        Console.WriteLine($"Output: {outputPath}");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("\nParameters:");
        Console.WriteLine(options.ToJson());
        Console.WriteLine("\nRunning evaluation...");

        EvaluationMetrics? metrics = null;
        string? error = null;
        bool success;

        try
        {
            metrics = await RunPerformanceEvaluation(options, outputPath);
            success = true;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"\nERROR: {ex.Message}");
            Console.Error.WriteLine(ex);
            error = ex.ToString();
            success = false;
        }
        finally
        {
            Console.Out.Flush();
            Console.SetOut(originalOut);
        }

        var finishedAt = DateTime.Now;
        var durationSeconds = (finishedAt - startedAt).TotalSeconds;

        var reportData = new ReportData
        {
            RunId = runId,
            StartedAt = startedAt,
            FinishedAt = finishedAt,
            DurationSeconds = durationSeconds,
            Success = success,
            Error = error,
            Parameters = options.ToParameters(),
            Environment = EnvironmentInfo.Create(),
            Metrics = metrics
        };

        var jsonOptions = new JsonSerializerOptions
        {
            WriteIndented = true,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };

        var jsonPath = Path.Combine(outputPath, "report.json");
        File.WriteAllText(jsonPath, JsonSerializer.Serialize(reportData, jsonOptions));
        Console.WriteLine($"\n✅ Saved: {jsonPath}");

        if (success && metrics is not null)
        {
            var mdPath = Path.Combine(outputPath, "report.md");
            File.WriteAllText(mdPath, GenerateMarkdownReport(reportData));
            Console.WriteLine($"✅ Saved: {mdPath}");
        }

        var logPath = Path.Combine(outputPath, "stdout.log");
        var logContent = teeWriter.CapturedText + (error is not null ? $"\nERROR:\n{error}" : string.Empty);
        File.WriteAllText(logPath, logContent);
        Console.WriteLine($"✅ Saved: {logPath}");

        Console.WriteLine("\n" + new string('=', 60));
        Console.WriteLine("EVALUATION COMPLETE");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine($"Duration: {durationSeconds:F2}s");
        Console.WriteLine($"Reports saved to: {outputPath}");

        return success ? 0 : 1;
    }

    private static async Task<EvaluationMetrics> RunPerformanceEvaluation(EvalOptions options, string outputPath)
    {
        var beforeDbPath = Path.Combine(outputPath, "crm_contacts_before.sqlite");
        var afterDbPath = Path.Combine(outputPath, "crm_contacts_after.sqlite");

        EvaluationMetrics metrics;

        await using (var beforeContext = new RepoBefore.CrmContactSearch.CrmDbContext(beforeDbPath, deleteDatabaseOnDispose: false))
        await using (var afterContext = new RepoAfter.CrmContactSearch.CrmDbContext(afterDbPath))
        {
            await beforeContext.EnsureCreatedAndSeededAsync();
            await afterContext.EnsureCreatedAndSeededAsync();

            var seedStats = await BuildSeedStats(beforeContext, afterContext);
            var scenarios = BuildScenarios(options);

            var beforeSearcher = new RepoBefore.CrmContactSearch.ContactSearcher(beforeContext);
            var afterSearcher = new RepoAfter.CrmContactSearch.ContactSearcher(afterContext);

            metrics = new EvaluationMetrics
            {
                SeedStats = seedStats
            };

            foreach (var scenario in scenarios)
            {
                var scenarioMetrics = await MeasureScenarioAsync(scenario, beforeSearcher, afterSearcher, options.Iterations);
                metrics.Before[scenario.Name] = scenarioMetrics.Before;
                metrics.After[scenario.Name] = scenarioMetrics.After;
                metrics.Comparison[scenario.Name] = scenarioMetrics.Comparison;

                Console.WriteLine($"\nScenario: {scenario.Name} ({scenario.Description})");
                Console.WriteLine($"  Before: {scenarioMetrics.Before.AvgMs:F2} ms avg " +
                                  $"(min {scenarioMetrics.Before.MinMs:F2}, max {scenarioMetrics.Before.MaxMs:F2})");
                Console.WriteLine($"  After : {scenarioMetrics.After.AvgMs:F2} ms avg " +
                                  $"(min {scenarioMetrics.After.MinMs:F2}, max {scenarioMetrics.After.MaxMs:F2})");
                Console.WriteLine($"  Speedup: {scenarioMetrics.Comparison.Speedup:F2}x");
            }

            metrics.Summary = BuildSummary(metrics);
        }

        CleanupDatabaseArtifacts(beforeDbPath);
        CleanupDatabaseArtifacts(afterDbPath);

        return metrics;
    }

    private static List<QueryScenario> BuildScenarios(EvalOptions options)
    {
        var pageSize = Math.Clamp(options.FilesPerFolder, 1, 200);

        return new List<QueryScenario>
        {
            new(
                "baseline",
                "Unfiltered search with minimum deal value",
                new RequestSpec(null, null, null, Stage.Qualified, 10_000m, 1, pageSize)
            ),
            new(
                "city_and_stage",
                "City + deal stage filter",
                new RequestSpec("London", null, null, Stage.Negotiation, 20_000m, 1, pageSize)
            ),
            new(
                "tags_and_value",
                "Tag filter plus higher deal value",
                new RequestSpec(null, "VIP,Premium", null, null, 50_000m, 1, pageSize)
            ),
            new(
                "stale_contacts_page_2",
                "Older contacts on page 2 with light floor",
                new RequestSpec(null, null, DateTime.UtcNow.AddDays(-120), Stage.Prospect, 5_000m, 2, pageSize)
            )
        };
    }

    private static async Task<ScenarioMetrics> MeasureScenarioAsync(
        QueryScenario scenario,
        RepoBefore.CrmContactSearch.ContactSearcher beforeSearcher,
        RepoAfter.CrmContactSearch.ContactSearcher afterSearcher,
        int iterations)
    {
        var beforeTimings = new List<double>(iterations);
        var afterTimings = new List<double>(iterations);
        SearchSnapshot? beforeSnapshot = null;
        SearchSnapshot? afterSnapshot = null;

        for (var i = 0; i < iterations; i++)
        {
            var result = await beforeSearcher.SearchContacts(ToBeforeRequest(scenario.Request));
            beforeTimings.Add(result.ElapsedMilliseconds);
            beforeSnapshot = new SearchSnapshot(result.TotalCount, result.Data.Count, result.Page, result.PageSize);
        }

        for (var i = 0; i < iterations; i++)
        {
            var result = await afterSearcher.SearchContacts(ToAfterRequest(scenario.Request));
            afterTimings.Add(result.ElapsedMilliseconds);
            afterSnapshot = new SearchSnapshot(result.TotalCount, result.Data.Count, result.Page, result.PageSize);
        }

        var beforeStats = TimingStats.From(beforeTimings, beforeSnapshot);
        var afterStats = TimingStats.From(afterTimings, afterSnapshot);
        var comparison = ComparisonStats.From(beforeStats.AvgMs, afterStats.AvgMs);

        return new ScenarioMetrics(scenario.Name, scenario.Description, beforeStats, afterStats, comparison);
    }

    private static RepoBefore.CrmContactSearch.ContactSearchRequest ToBeforeRequest(RequestSpec request)
    {
        return new RepoBefore.CrmContactSearch.ContactSearchRequest
        {
            City = request.City,
            Tags = request.Tags,
            LastContactBefore = request.LastContactBefore,
            DealStage = request.DealStage is null ? null : (RepoBefore.CrmContactSearch.DealStage?)request.DealStage,
            MinDealValue = request.MinDealValue,
            Page = request.Page,
            PageSize = request.PageSize
        };
    }

    private static RepoAfter.CrmContactSearch.ContactSearchRequest ToAfterRequest(RequestSpec request)
    {
        return new RepoAfter.CrmContactSearch.ContactSearchRequest
        {
            City = request.City,
            Tags = request.Tags,
            LastContactBefore = request.LastContactBefore,
            DealStage = request.DealStage is null ? null : (RepoAfter.CrmContactSearch.DealStage?)request.DealStage,
            MinDealValue = request.MinDealValue,
            Page = request.Page,
            PageSize = request.PageSize
        };
    }

    private static SummaryMetrics BuildSummary(EvaluationMetrics metrics)
    {
        double beforeAvg = metrics.Before.Count == 0 ? 0 : metrics.Before.Values.Average(m => m.AvgMs);
        double afterAvg = metrics.After.Count == 0 ? 0 : metrics.After.Values.Average(m => m.AvgMs);
        double overallSpeedup = afterAvg > 0 ? beforeAvg / afterAvg : 0;

        return new SummaryMetrics
        {
            BeforeAvgMs = Math.Round(beforeAvg, 2),
            AfterAvgMs = Math.Round(afterAvg, 2),
            OverallSpeedup = Math.Round(overallSpeedup, 2),
            OverallImprovementPct = beforeAvg > 0
                ? Math.Round(((beforeAvg - afterAvg) / beforeAvg) * 100, 1)
                : 0
        };
    }

    private static async Task<SeedStats> BuildSeedStats(
        RepoBefore.CrmContactSearch.CrmDbContext beforeContext,
        RepoAfter.CrmContactSearch.CrmDbContext afterContext)
    {
        return new SeedStats
        {
            Contacts = await beforeContext.Contacts.CountAsync(),
            Tags = await beforeContext.Tags.CountAsync(),
            Interactions = await beforeContext.Interactions.CountAsync(),
            Deals = await beforeContext.Deals.CountAsync(),
            CachedMetrics = await afterContext.ContactMetrics.CountAsync()
        };
    }

    private static void CleanupDatabaseArtifacts(string databasePath)
    {
        DeleteIfExists(databasePath);
        DeleteIfExists(databasePath + "-wal");
        DeleteIfExists(databasePath + "-shm");
    }

    private static void DeleteIfExists(string path)
    {
        if (!File.Exists(path))
        {
            return;
        }

        try
        {
            File.Delete(path);
        }
        catch (IOException)
        {
            // Best-effort cleanup only
        }
        catch (UnauthorizedAccessException)
        {
            // Best-effort cleanup only
        }
    }

    private static string GenerateMarkdownReport(ReportData report)
    {
        if (report.Metrics is null)
        {
            return "# Performance Evaluation Report\n\nRun did not complete.";
        }

        var env = report.Environment;
        var metrics = report.Metrics;
        var parameters = report.Parameters;

        var sb = new StringBuilder();
        sb.AppendLine("# Performance Evaluation Report");
        sb.AppendLine();
        sb.AppendLine($"**Run ID:** `{report.RunId}`  ");
        sb.AppendLine($"**Started:** {report.StartedAt:O}  ");
        sb.AppendLine($"**Finished:** {report.FinishedAt:O}  ");
        sb.AppendLine($"**Duration:** {report.DurationSeconds:F2} seconds");
        sb.AppendLine("\n---\n");

        sb.AppendLine("## Environment\n");
        sb.AppendLine("| Property | Value |");
        sb.AppendLine("|----------|-------|");
        sb.AppendLine($"| .NET | {env.DotNetVersion} |");
        sb.AppendLine($"| Runtime | {env.FrameworkDescription} |");
        sb.AppendLine($"| OS | {env.OsDescription} |");
        sb.AppendLine($"| Architecture | {env.Architecture} |");
        sb.AppendLine($"| Processors | {env.ProcessorCount} |");
        sb.AppendLine($"| Git Commit | `{env.GitCommit}` |");
        sb.AppendLine($"| Git Branch | `{env.GitBranch}` |");
        sb.AppendLine("\n---\n");

        sb.AppendLine("## Parameters\n");
        sb.AppendLine("| Parameter | Value |");
        sb.AppendLine("|-----------|-------|");
        sb.AppendLine($"| Users | {parameters.Users} |");
        sb.AppendLine($"| Folders | {parameters.Folders} |");
        sb.AppendLine($"| Files per Folder | {parameters.FilesPerFolder} |");
        sb.AppendLine($"| Iterations | {parameters.Iterations} |");
        sb.AppendLine("\n---\n");

        sb.AppendLine("## Seed Statistics\n");
        sb.AppendLine("| Contacts | Tags | Interactions | Deals | Cached Metrics |");
        sb.AppendLine("|----------|------|--------------|-------|----------------|");
        sb.AppendLine($"| {metrics.SeedStats.Contacts} | {metrics.SeedStats.Tags} | {metrics.SeedStats.Interactions} | {metrics.SeedStats.Deals} | {metrics.SeedStats.CachedMetrics} |");
        sb.AppendLine("\n---\n");

        sb.AppendLine("## Summary Results\n");
        sb.AppendLine("| Metric | Before (Naive) | After (Optimized) | Improvement |");
        sb.AppendLine("|--------|----------------|-------------------|-------------|");
        sb.AppendLine($"| Average Response | {metrics.Summary.BeforeAvgMs:F2} ms | {metrics.Summary.AfterAvgMs:F2} ms | **{metrics.Summary.OverallSpeedup:F2}x faster** |");
        sb.AppendLine($"| Improvement | - | - | {metrics.Summary.OverallImprovementPct:F1}% |");
        sb.AppendLine("\n---\n");

        sb.AppendLine("## Detailed Results by Scenario\n");
        sb.AppendLine("| Scenario | Before (ms) | After (ms) | Speedup | Returned | Total |");
        sb.AppendLine("|----------|-------------|------------|---------|----------|-------|");

        foreach (var scenario in metrics.Before.Keys)
        {
            var b = metrics.Before[scenario];
            var a = metrics.After[scenario];
            var c = metrics.Comparison[scenario];
            sb.AppendLine($"| {scenario} | {b.AvgMs:F2} | {a.AvgMs:F2} | {c.Speedup:F2}x | {a.Returned} | {a.Total} |");
        }

        sb.AppendLine("\n---\n");
        sb.AppendLine("## Conclusion\n");

        if (metrics.Summary.OverallSpeedup >= 2.0)
        {
            sb.AppendLine($"✅ **Excellent optimization.** The optimized implementation is **{metrics.Summary.OverallSpeedup:F2}x faster** than the naive version.");
        }
        else if (metrics.Summary.OverallSpeedup >= 1.0)
        {
            sb.AppendLine($"✅ **Good optimization.** The optimized implementation is **{metrics.Summary.OverallSpeedup:F2}x faster** than the naive version.");
        }
        else
        {
            sb.AppendLine("⚠️ **Needs investigation.** The optimized implementation is slower than the naive version.");
        }

        return sb.ToString();
    }
}

internal sealed record EvalOptions
{
    public int Users { get; init; } = 21;
    public int Folders { get; init; } = 100;
    public int FilesPerFolder { get; init; } = 50;
    public int Iterations { get; init; } = 5;
    public string OutputDir { get; init; } = "evaluation";

    public static EvalOptions Parse(string[] args)
    {
        var options = new EvalOptions();

        for (var i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "--users" when i + 1 < args.Length && int.TryParse(args[i + 1], out var users):
                    options = options with { Users = users };
                    i++;
                    break;
                case "--folders" when i + 1 < args.Length && int.TryParse(args[i + 1], out var folders):
                    options = options with { Folders = folders };
                    i++;
                    break;
                case "--files-per-folder" when i + 1 < args.Length && int.TryParse(args[i + 1], out var fpf):
                    options = options with { FilesPerFolder = fpf };
                    i++;
                    break;
                case "--iterations" when i + 1 < args.Length && int.TryParse(args[i + 1], out var iterations):
                    options = options with { Iterations = iterations };
                    i++;
                    break;
                case "--output-dir" when i + 1 < args.Length:
                    options = options with { OutputDir = args[i + 1] };
                    i++;
                    break;
            }
        }

        return options;
    }

    public string ToJson()
    {
        var opts = new JsonSerializerOptions { WriteIndented = true };
        return JsonSerializer.Serialize(ToParameters(), opts);
    }

    public EvalParameters ToParameters() =>
        new()
        {
            Users = Users,
            Folders = Folders,
            FilesPerFolder = FilesPerFolder,
            Iterations = Iterations
        };
}

internal sealed class EvalParameters
{
    public int Users { get; init; }
    public int Folders { get; init; }
    public int FilesPerFolder { get; init; }
    public int Iterations { get; init; }
}

internal sealed class EnvironmentInfo
{
    public string DotNetVersion { get; init; } = string.Empty;
    public string FrameworkDescription { get; init; } = string.Empty;
    public string OsDescription { get; init; } = string.Empty;
    public string Architecture { get; init; } = string.Empty;
    public int ProcessorCount { get; init; }
    public string GitCommit { get; init; } = "unknown";
    public string GitBranch { get; init; } = "unknown";

    public static EnvironmentInfo Create()
    {
        return new EnvironmentInfo
        {
            DotNetVersion = Environment.Version.ToString(),
            FrameworkDescription = RuntimeInformation.FrameworkDescription,
            OsDescription = RuntimeInformation.OSDescription,
            Architecture = RuntimeInformation.OSArchitecture.ToString(),
            ProcessorCount = Environment.ProcessorCount,
            GitCommit = TryGit("rev-parse --short HEAD"),
            GitBranch = TryGit("rev-parse --abbrev-ref HEAD")
        };
    }

    private static string TryGit(string arguments)
    {
        try
        {
            var psi = new ProcessStartInfo("git", arguments)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using var process = Process.Start(psi);
            if (process is null)
            {
                return "unknown";
            }

            var output = process.StandardOutput.ReadToEnd().Trim();
            process.WaitForExit(3000);
            return process.ExitCode == 0 ? output : "unknown";
        }
        catch
        {
            return "unknown";
        }
    }
}

internal sealed class ReportData
{
    public string RunId { get; init; } = string.Empty;
    public DateTime StartedAt { get; init; }
    public DateTime FinishedAt { get; init; }
    public double DurationSeconds { get; init; }
    public bool Success { get; init; }
    public string? Error { get; init; }
    public EvalParameters Parameters { get; init; } = new();
    public EnvironmentInfo Environment { get; init; } = new();
    public EvaluationMetrics? Metrics { get; init; }
}

internal sealed class SeedStats
{
    public int Contacts { get; init; }
    public int Tags { get; init; }
    public int Interactions { get; init; }
    public int Deals { get; init; }
    public int CachedMetrics { get; init; }
}

internal sealed class TimingStats
{
    public double AvgMs { get; init; }
    public double MinMs { get; init; }
    public double MaxMs { get; init; }
    public int Total { get; init; }
    public int Returned { get; init; }
    public int Page { get; init; }
    public int PageSize { get; init; }

    public static TimingStats From(IReadOnlyCollection<double> timings, SearchSnapshot? snapshot)
    {
        if (timings.Count == 0)
        {
            return new TimingStats();
        }

        return new TimingStats
        {
            AvgMs = Math.Round(timings.Average(), 2),
            MinMs = Math.Round(timings.Min(), 2),
            MaxMs = Math.Round(timings.Max(), 2),
            Total = snapshot?.Total ?? 0,
            Returned = snapshot?.Returned ?? 0,
            Page = snapshot?.Page ?? 0,
            PageSize = snapshot?.PageSize ?? 0
        };
    }
}

internal sealed class ComparisonStats
{
    public double Speedup { get; init; }
    public double ImprovementPct { get; init; }

    public static ComparisonStats From(double beforeAvg, double afterAvg)
    {
        var speedup = afterAvg > 0 ? beforeAvg / afterAvg : 0;
        var improvement = beforeAvg > 0
            ? ((beforeAvg - afterAvg) / beforeAvg) * 100
            : 0;

        return new ComparisonStats
        {
            Speedup = Math.Round(speedup, 2),
            ImprovementPct = Math.Round(improvement, 1)
        };
    }
}

internal sealed class SummaryMetrics
{
    public double BeforeAvgMs { get; init; }
    public double AfterAvgMs { get; init; }
    public double OverallSpeedup { get; init; }
    public double OverallImprovementPct { get; init; }
}

internal sealed class EvaluationMetrics
{
    public SeedStats SeedStats { get; init; } = new();
    public Dictionary<string, TimingStats> Before { get; } = new(StringComparer.OrdinalIgnoreCase);
    public Dictionary<string, TimingStats> After { get; } = new(StringComparer.OrdinalIgnoreCase);
    public Dictionary<string, ComparisonStats> Comparison { get; } = new(StringComparer.OrdinalIgnoreCase);
    public SummaryMetrics Summary { get; set; } = new();
}

internal sealed record QueryScenario(string Name, string Description, RequestSpec Request);

internal sealed record RequestSpec(
    string? City,
    string? Tags,
    DateTime? LastContactBefore,
    Stage? DealStage,
    decimal? MinDealValue,
    int Page,
    int PageSize
);

internal sealed record ScenarioMetrics(
    string Name,
    string Description,
    TimingStats Before,
    TimingStats After,
    ComparisonStats Comparison
);

internal sealed record SearchSnapshot(int Total, int Returned, int Page, int PageSize);

internal enum Stage
{
    Prospect = 0,
    Qualified = 1,
    Proposal = 2,
    Negotiation = 3,
    ClosedWon = 4,
    ClosedLost = 5
}

internal sealed class TeeTextWriter : TextWriter
{
    private readonly TextWriter _inner;
    private readonly StringBuilder _buffer = new();

    public TeeTextWriter(TextWriter inner)
    {
        _inner = inner;
    }

    public string CapturedText => _buffer.ToString();

    public override Encoding Encoding => _inner.Encoding;

    public override void Write(char value)
    {
        _inner.Write(value);
        _buffer.Append(value);
    }

    public override void Write(string? value)
    {
        _inner.Write(value);
        _buffer.Append(value);
    }

    public override void WriteLine(string? value)
    {
        _inner.WriteLine(value);
        _buffer.AppendLine(value);
    }
}
